In [277]:
from IPython.display import HTML
import plotly.io as pio
import warnings
warnings.filterwarnings("ignore")
pio.renderers.default = "notebook_connected"
HTML('''<button type="button" class="btn btn-outline-danger"  onclick="codeToggle();">Toggle Code</button>''')


In [278]:
import pandas as pd

In [309]:
consumption_full = pd.read_csv("/Users/na/campus_energy_writing/consumption.csv")

In [279]:
q1 = pd.read_csv("/Users/na/campus_energy_writing/q1.csv")
q2 = pd.read_csv("/Users/na/campus_energy_writing/q2.csv")
q3 = pd.read_csv("/Users/na/campus_energy_writing/q3.csv")
q4 = pd.read_csv("/Users/na/campus_energy_writing/q4.csv")

In [280]:
chiller_q1 = pd.read_csv("/Users/na/campus_energy_writing/chiller_q1.csv")
chiller_q2 = pd.read_csv("/Users/na/campus_energy_writing/chiller_q2.csv")
chiller_q3 = pd.read_csv("/Users/na/campus_energy_writing/chiller_q3.csv")
chiller_q4 = pd.read_csv("/Users/na/campus_energy_writing/chiller_q4.csv")

In [281]:
weather = pd.read_csv("/Users/na/campus_energy_writing/weather.csv")

In [282]:
humidity = weather[weather['Metric']=="Humidity"]

In [283]:
humidity = list(weather[weather['Metric']=="Humidity"].iloc[0])
humidity[1:]

[43.0, 39.0, 31.0, 31.0, 42.0, 58.0, 76.0, 80.0, 74.0, 52.0, 44.0, 46.0]

In [284]:
temp = list(weather[weather['Metric']=="Avg Temp"].iloc[0])
temp = temp[1:]

In [285]:
sunny_day_data = list(weather[weather['Metric']=="avg sun hours"].iloc[0])

In [286]:
longer_supply = pd.read_csv("/Users/na/campus_energy_writing/supply.csv")

solar_21 = longer_supply[longer_supply['Year']==21]['Solar Power']
solar_22 = longer_supply[longer_supply['Year']==22]['Solar Power']
temp21 = temp
temp22 = temp


<h2>Primary Hypothesis</h2>
<p>
  <strong>
    IIT Gandhinagar's campus uses the highest supply of energy from Torrent
    Power in October, and solar power is the best cure to address this
    surge.
  </strong>
</p>

In [287]:
import pandas as pd
import plotly.express as px

# Data extracted from the dataset
data = {
    'Month': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 
              'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'] * 2,
    'Year': [21] * 12 + [22] * 12,
    'Torrent Power':list(longer_supply['Torrent Power'])
}

# Create a DataFrame
df = pd.DataFrame(data)

# Create the line chart with slider
fig = px.line(
    df,
    x='Month',
    y='Torrent Power',
    color='Year',
    title="Torrent Power Consumption Across Months (2021 vs. 2022)",
    labels={'Torrent Power': 'Energy Consumption (kWh)', 'Month': 'Month'},
    animation_frame='Year',
    height=900
)

# Update layout for better readability
fig.update_layout(
    xaxis_title="Month",
    yaxis_title="Energy Consumption (kWh)",
    legend_title="Year",
    template="plotly_white",
    
)

# Show the plot
fig.show()


In [295]:
sunny2 = sunny_day_data[1:]*2
for month in range(len(sunny2)):
    sunny2[month] = 1.5**(sunny2[month])

In [297]:
combined_data = {
    'Month': chiller_data['Month'],
    'Solar': solar_data,
    'Chiller': chiller_data['Consumption']
}

fig = px.line(combined_data, x='Month', y=['Solar','Chiller'],
             title="Solar Generation vs Chiller Demand",
             labels={'value':'kWh'})
fig.update_layout(yaxis2=dict(title='Chiller (kWh)', overlaying='y', side='right'))
fig.show()


In [298]:
weather

Metric  January  February  March  April   May  June   July  August  \
0       Avg Temp     20.4      22.8   27.2   31.2  33.0  31.9   28.5    27.5   
1  Precipitation      1.0       1.0    1.0    1.0   1.0  73.0  307.0   242.0   
2       Humidity     43.0      39.0   31.0   31.0  42.0  58.0   76.0    80.0   
3     Rainy days      0.0       0.0    0.0    0.0   0.0   5.0   14.0    15.0   
4  avg sun hours      9.7      10.2   10.8   11.4  11.3   9.5    7.1     6.2   

   September  October  November  December  
0       28.0     27.9      25.0      21.5  
1      109.0     17.0       3.0       1.0  
2       74.0     52.0      44.0      46.0  
3        7.0      2.0       1.0       0.0  
4        8.0     10.1       9.9       9.6

In [299]:
import numpy as np

<h2>Sub-Hypotheses</h2>


<h3>1. October Experiences Maximum Transmission Loss</h3>
<p>
  Transmission loss increases tremendously in October as the campus grid
  is heavily loaded by different activities. Though the chiller power
  consumption slightly drops due to favorable weather, this savings is
  overpowered by the increased demand from academic and sports facilities.
</p>

In [300]:
temperatures = weather[weather['Metric']=='Avg Temp'].iloc[0]
temperatures = list(temperatures)
temperatures = temperatures[1:]
temperatures

[20.4, 22.8, 27.2, 31.2, 33.0, 31.9, 28.5, 27.5, 28.0, 27.9, 25.0, 21.5]

In [301]:
q1_t = np.mean(temperatures[0:3])
q2_t = np.mean(temperatures[3:6])
q3_t = np.mean(temperatures[6:9])
q4_t = np.mean(temperatures[9:12])

In [302]:
loss_data = {
    'Quarter': ['Q1','Q2','Q3','Q4'],
    'Loss%': [1.68,1.31,1.70,3.31],
    'Temperature': [q1_t,q2_t,q3_t,q4_t]
}

fig = px.bar(loss_data, x='Quarter', y='Loss%',
            title="Transmission Losses by Quarter",
            color='Temperature')
fig.update_traces(texttemplate='%{y}%', textposition='outside')


In [303]:
def make_convertible(s):
    s = s.split(",")
    s = "".join(s)
    return s

<h3>2. Increased Academic Block Power Usage in October</h3>
<p>
  During October, the academic calendar peaks with exams and related
  activities, pushing the energy consumption in academic buildings to
  higher levels. Key contributing factors include:
</p>
<ul>
  <li>Extended operating hours for classrooms, libraries, and laboratories.</li>
  <li>
    Enhanced use of air-conditioning, lighting, and electronic devices during
    exam times.
  </li>
  <li>Additional support for multimedia setups, projectors, and computers.</li>
</ul>


In [304]:
import plotly.graph_objects as go

categories = ['Q1','Q2','Q3','Q4']
academic = [int(make_convertible(q1.iloc[3]['Academic Area'])),int(make_convertible(q2.iloc[3]['Academic Area'])),int(make_convertible(q3.iloc[3]['Academic Area'])),int(make_convertible(q4.iloc[3]['Academic Area']))]
hostel = [int(make_convertible(q1.iloc[3]['Hostels & Guest house'])),int(make_convertible(q2.iloc[3]['Hostels & Guest house'])),int(make_convertible(q3.iloc[3]['Hostels & Guest house'])),int(make_convertible(q4.iloc[3]['Hostels & Guest house']))]
sports = [int(make_convertible(q1.iloc[3]['Sports Complex'])),int(make_convertible(q2.iloc[3]['Sports Complex'])),int(make_convertible(q3.iloc[3]['Sports Complex'])),int(make_convertible(q4.iloc[3]['Sports Complex']))]

fig = go.Figure(data=[
    go.Bar(name='Academic', x=categories, y=academic),
    go.Bar(name='Hostel', x=categories, y=hostel)
])
fig.update_layout(barmode='stack', title="Energy Use: Academic vs Residential Areas")



<h3>3. Increased Power Usage in the Sports Complex</h3>
<p>
  In preparation for the Inter-IIT Sports Meet, the sports complex faces a
  spike in energy consumption driven by:
</p>
<ul>
  <li>High usage of floodlights during evening practice sessions.</li>
  <li>Extended operation of gym and training facilities.</li>
  <li>Increased demand for water heating in locker rooms and related amenities.</li>
</ul>

In [307]:
fig = go.Figure()

# Add Academic Area data
fig.add_trace(go.Scatter(
    x=categories,
    y=academic,
    mode='lines+markers',
    name='Academic Area',
    line=dict(color='blue'),
    marker=dict(size=8)
))

# Add Hostel data
fig.add_trace(go.Scatter(
    x=categories,
    y=hostel,
    mode='lines+markers',
    name='Hostels & Guest House',
    line=dict(color='red'),
    marker=dict(size=8),
    
))

# Update layout for better readability
fig.update_layout(
    title="Comparison of Energy Consumption: Academic vs Hostels",
    xaxis_title="Quarter",
    yaxis_title="Energy Consumption (kWh)",
    legend_title="Area",
    template="plotly_white",
    
)

# Show the plot
fig.show()

<h3>4. Reduced Chiller Power Consumption in October</h3>
<p>
  The chiller systems show a decline in energy consumption in October
  because of:
</p>
<ul>
  <li>Moderate average temperatures (approximately 27.9°C).</li>
  <li>Lower humidity levels (around 52%) compared to the monsoon peaks.</li>
  <li>
    Fewer rainy days, which reduce the latent heat load typically managed by
    chillers.
  </li>
</ul>



In [288]:
import plotly.express as px

chiller_data = {
    'Month': ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'],
    'Consumption': list(chiller_q1['Consumption of Power of Chiller Plant'][:3])+list(chiller_q2['Consumption of Power of Chiller Plant'][:3])+list(chiller_q3['Consumption of Power of Chiller Plant'][:3])+list(chiller_q4['Consumption of Power of Chiller Plant'][:3])
}

fig = px.area(chiller_data, x='Month', y='Consumption',
             title="Chiller Plant Energy Consumption (2021)",
             labels={'Consumption':'kWh'})
fig.add_scatter(x=chiller_data['Month'], y=chiller_data['Consumption'], 
               mode='lines+markers', name='Actual')
fig.show()


<h3>
  Despite this reduction, the combined rise in power demand from academic
  and sports activities leads to an overall increase in campus energy need,
  resulting in the highest supply requirements from Torrent Power during
  October.
</h3>

In [311]:
import pandas as pd
import plotly.express as px


# Transform the DataFrame from wide to long format; each row will represent a Month–Category pair
df_long = consumption_full.melt(id_vars="Month", 
                  var_name="Category", 
                  value_name="Consumption")

# Create an animated horizontal bar chart (race chart) using Plotly Express
fig = px.bar(
    df_long,
    x="Consumption",
    y="Category",
    animation_frame="Month",
    orientation="h",
    range_x=[0, df_long["Consumption"].max() * 1.1],
    title="Power Consumption Race Chart",
    labels={"Consumption": "Consumption (Units)", "Category": "Category"}
)

# Optional: sort the bars based on consumption values in each frame
fig.update_layout(yaxis={'categoryorder': 'total ascending'})

fig.show()


<h1> Why especially in October, Increasing Solar Power Generation Capabilities minimized External Power dependency even more effectively? </h1>

<h3>Data Supporting Solar Efficiency in October:</h3>

In [289]:
solar_data = list(q1['Solar Power'][:3])+list(q2['Solar Power'][:3])+list(q3['Solar Power'][:3])+list(q4['Solar Power'][:3])

for i in range(len(solar_data)):
    solar_data[i]="".join(solar_data[i].split(","))
    solar_data[i] = float(solar_data[i])

In [290]:
import plotly.express as px

chiller_data = {
    'Month': ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'],
    'Consumption': solar_data,
}

fig = px.area(chiller_data, x='Month', y='Consumption',
             title="Solar Power Generation (2021)",
             labels={'Consumption':'kWh'})
fig.add_scatter(x=chiller_data['Month'], y=chiller_data['Consumption'], 
               mode='lines+markers', name='Actual')
fig.show()



<p>
  The following table presents the monthly environmental metrics for IIT
  Gandhinagar, highlighting why October stands out:
</p>

<table>
  <caption>Monthly Environmental Metrics at IIT Gandhinagar</caption>
  <thead>
    <tr>
      <th>Metric</th>
      <th>January</th>
      <th>February</th>
      <th>March</th>
      <th>April</th>
      <th>May</th>
      <th>June</th>
      <th>July</th>
      <th>August</th>
      <th>September</th>
      <th>October</th>
      <th>November</th>
      <th>December</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Avg Temp (°C)</td>
      <td>20.4</td>
      <td>22.8</td>
      <td>27.2</td>
      <td>31.2</td>
      <td>33.0</td>
      <td>31.9</td>
      <td>28.5</td>
      <td>27.5</td>
      <td>28.0</td>
      <td>27.9</td>
      <td>25.0</td>
      <td>21.5</td>
    </tr>
    <tr>
      <td>Precipitation (mm)</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>73.0</td>
      <td>307.0</td>
      <td>242.0</td>
      <td>109.0</td>
      <td>17.0</td>
      <td>3.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <td>Humidity (%)</td>
      <td>43.0</td>
      <td>39.0</td>
      <td>31.0</td>
      <td>31.0</td>
      <td>42.0</td>
      <td>58.0</td>
      <td>76.0</td>
      <td>80.0</td>
      <td>74.0</td>
      <td>52.0</td>
      <td>44.0</td>
      <td>46.0</td>
    </tr>
    <tr>
      <td>Rainy days</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>5</td>
      <td>14</td>
      <td>15</td>
      <td>7</td>
      <td>2</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <td>Avg Sun Hours</td>
      <td>9.7</td>
      <td>10.2</td>
      <td>10.8</td>
      <td>11.4</td>
      <td>11.3</td>
      <td>9.5</td>
      <td>7.1</td>
      <td>6.2</td>
      <td>8.0</td>
      <td>10.1</td>
      <td>9.9</td>
      <td>9.6</td>
    </tr>
  </tbody>
</table>

<h3>Optimal Conditions for Solar Efficiency in October</h3>
<p>
  October is particularly ideal for solar power production because:
</p>
<ul>
  <li>
    The average temperature of 27.9°C is moderate enough to minimize thermal
    losses, which can hamper solar panel efficiency in excessively hot months.
  </li>
  <li>
    With an average of 10.1 sun hours per day in October, there is ample
    sunlight to drive significant solar power generation.
  </li>
  <li>
    Low precipitation and minimal rainy days reduce cloud cover, ensuring more
    consistent solar exposure.
  </li>
</ul>


In [291]:
solar_data_ = solar_data
for i in range(len(solar_data)):
    solar_data_[i]/= 19000

<p>
  Additional data visualizations, such as a temperature vs. power graph and a
  bubble chart (with the x-axis as temperature, y-axis as power output in kWh,
  and bubble size reflecting the number of sunny days), further validate that
  at optimal range even with less sunny days, solar panels are able to generate decent solar energy.
</p>

In [292]:
import plotly.express as px

plot_data = {
    'Month': ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'],
    'sunny': sunny_day_data[1:],
    'temp': temp,
    'solar': solar_data_
}

fig = px.scatter(plot_data, x='temp', y='sunny', size='solar',
                hover_name='Month', title='Solar Generation vs Sunny Days vs Temperature',
                labels={'Chiller':'Chiller Load (kWh)', 'Solar':'Solar Generation (kWh)'})
fig.show()


In [293]:
import numpy as np

In [294]:
temp

[20.4, 22.8, 27.2, 31.2, 33.0, 31.9, 28.5, 27.5, 28.0, 27.9, 25.0, 21.5]

<p>
  Adding to it, the below graph proves that at very high temperatures, despite having many sunny days, solar panels are not able to generate much energy.
</p>

In [296]:
import plotly.express as px
import pandas as pd

# Sample data (replace with your actual data)
# solar_21 = [53492,55250,72111,66687,59496,58152,45676,46142,39108,59193,55470,46258]
# solar_22 = [60000,62000,75000,70000,65000,63000,50000,51000,42000,61000,57000,48000]
# temp = [15,16,20,25,30,32,35,34,28,22,18,16]  # Single temperature array for both years

# Create structured DataFrame
data = {
    'Month': ['Jan','Feb','Mar','Apr','May','Jun',
             'Jul','Aug','Sep','Oct','Nov','Dec'] * 2,
    'Solar Power': list(solar_21) + list(solar_22),
    'Temperature': temp * 2,  # Reuse same temp pattern for both years
    'Sunny Days': sunny2,
    'Year': [2021]*12 + [2022]*12
}

df = pd.DataFrame(data)

# Create animated plot
fig = px.scatter(
    df, 
    x='Temperature', 
    y='Solar Power', 
    color='Year',
    size='Sunny Days',
    animation_frame='Year',
    title='Solar Power vs Temperature (2021 vs 2022)',
    labels={'Temperature': 'Temperature (°C)', 'Solar Power': 'kWh'},
    range_x=[17,37], 
    range_y=[30000,80000],
    height=1000
)
fig.update_layout(
    plot_bgcolor="rgba(255,255,255)", # Transparent background
    paper_bgcolor="rgba(255,255,255)", # Transparent paper background
    shapes=[
        # Add a rectangle to indicate the temperature range (19°C to 33°C)
        dict(
            type="rect",
            xref="x",
            yref="paper",
            x0=17,
            x1=31,
            y0=0,
            y1=1,
            fillcolor="rgba(173,216,230,0.8)", # Light blue background for range
            layer="below",
            line_width=0,
        )
    ]
)
# fig.update_layout(sliders_orientation="horizontal")
fig.show()
